### PR TITLE
Scoping email

### PR DESCRIPTION
**Pull request by vinyll**

### CODE REVIEW

File: core/settings/base.py [hunk 1]
```diff
@@ -48,6 +48,7 @@
 )
 
 AUTHENTICATION_BACKENDS = (
+    'django.contrib.auth.backends.ModelBackend',
```
Comment (vinyll) [inline core/settings/base.py:51]: This had been deactivated which disabled the regular email/password authentication!

File: api/tests.py [hunk 2]
```diff
@@ -7,31 +7,38 @@
 
 class ProfileTest(BaseTest):
 
+    def createToken(self, **kwargs):
+        defaults = {
+            'user': self.test_user, 'token': '1234567890', 'scope': 'read',
+            'application': self.application,
+            'expires': timezone.now() + datetime.timedelta(days=1),
+        }
+        kwargs = {**defaults, **kwargs}
```
Comment (davidbgk) [inline api/tests.py]: `defaults.update(kwargs)` looks more familiar to me.

File: api/views.py [hunk 3]
```diff
@@ -1,18 +1,19 @@
...
     def get(self, request, *args, **kwargs):
         user = request.user
-        return JsonResponse({
-            'id': user.pk,
-            'email': user.email,
-        })
+        _, request = self.verify_request(request)
+        response = {'id': user.pk}
+
+        if request.access_token.scope == 'email':
```
Comment (davidbgk) [inline api/views.py:16]: I'm not sure where we want to go here in terms of abstraction/customization. I guess that this is not the default behavior that we should push upstream, still it's interesting to be able to explicitly define somewhere (settings?) the relations scopes ⟷ returned fields. We can do that in another PR but it's something we should not forget. Thoughts?

File: oauth/middleware.py [hunk 4]
```diff
@@ -19,8 +19,7 @@ def __call__(self, request):
                     request.user = request._cached_user = user
                 else:
                     response = HttpResponse('Unauthorized', status=401)
-                    response['WWW-Authenticate'] = (
-                        'Bearer realm="api"')
+                    response['WWW-Authenticate'] = ('Bearer realm="api"')
```
Comment (davidbgk) [inline oauth/middleware.py]: Are the parenthesis still required?

File: oauth/templates/oauth2_provider/authorize.html [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+                {% if scopes_descriptions %}
+                  <p>{% trans "Application requires following permissions:" %}</p>
+                  <ul>
+                      {% for scope in scopes_descriptions %}
+                          <li>{{ scope }}</li>
+                      {% endfor %}
+                  </ul>
+                {% else %}
+                  <p>The application does not require any specific permission.</p>
```
Comment (davidbgk) [inline oauth/templates/oauth2_provider/authorize.html:25]: Missing translation.

File: oauth/templates/oauth2_provider/authorize.html [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+                {% endif %}
+
+                {{ form.errors }}
+                {{ form.non_field_errors }}
+
+                <div class="control-group">
+                    <div class="controls">
+                        <input type="submit" class="btn btn-large" value="Cancel"/>
```
Comment (davidbgk) [inline oauth/templates/oauth2_provider/authorize.html:33]: Missing translation.

File: oauth/templates/oauth2_provider/authorize.html [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+                        <input type="submit" class="btn btn-large btn-primary" name="allow" value="Authorize"/>
```
Comment (davidbgk) [inline oauth/templates/oauth2_provider/authorize.html:34]: Missing translation.

File: oauth/templates/oauth2_provider/authorize.html [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+                    </div>
+                </div>
+            </form>
+
+        {% else %}
+            <h2>Error: {{ error.error }}</h2>
```
Comment (davidbgk) [inline oauth/templates/oauth2_provider/authorize.html:40]: Missing translation.

File: core/settings/base.py [hunk 9]
```diff
@@ -144,5 +145,10 @@
 # Name of the URL where a user may login
 LOGIN_URL = 'auth_login'
 
-
 AUTH_USER_MODEL = 'register.User'
+
+OAUTH2_PROVIDER = {
```
Comment (davidbgk) [inline core/settings/base.py:150]: So I guess that's the setting we have missed during our pair session, kudos for finding it! 👍

File: core/settings/base.py [hunk 10]
```diff
@@ -144,5 +145,10 @@
...
+    'SCOPES': {'email': 'Viewing your email address.'},
+    'READ_SCOPE': ['email'],
```
Comment (davidbgk) [inline core/settings/base.py:152]: Isn't it possible to use these settings to filter the returned dict instead of doing it manually within the code? (see my previous comment)

**Review by davidbgk — CHANGES_REQUESTED** [review]

File: api/views.py (hunk 3 above)
Comment (vinyll) [reply api/views.py:16]: We will probably need another _scope_ for the groups authorities for datagouv.
However I'm not sure it will behave the same way and I wouldn't go for abstration while we don't need it.

File: oauth/templates/oauth2_provider/authorize.html (hunk 6 above)
Comment (vinyll) [reply oauth/templates/oauth2_provider/authorize.html:33]: All this code was only a copy paste from the native app's template.
I'm not taking care of the translations right now as it would mean collecting each place that is not translated in the apps.
However Ybon was making the same notification.

File: core/settings/base.py (hunk 9 above)
Comment (vinyll) [reply core/settings/base.py:150]: We had this already. The key thing is [here](https://github.com/etalab/id.data.gouv.fr/pull/21/files#diff-265524153427c5c3f0dd9ae087f16a4bR13), using the returned enriched request instead of the local request.

File: core/settings/base.py (hunk 10 above)
Comment (vinyll) [reply core/settings/base.py:152]: It would be possible for sure, but I'm really not into making something more generic – aka more complex and less explicit – while we don't have the need.

**Comment by vinyll** [comment]
Translation are a wider topic than just these phrases.
I opened [an issue about that](#22).

File: api/views.py (hunk 3 above)
Comment (noirbizarre) [reply api/views.py:16]: Agreed with @davidbgk : Having one place to declare scopes ⟷fields relationships would easier to maintain/understand... But also agree that it should be another PR ;)

**Review by noirbizarre — APPROVED** [review]

File: api/views.py (hunk 3 above)
Comment (vinyll) [reply api/views.py:16]: https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it
Ok to implement it when we will actually have multiple scopes contexts.

**Comment by vinyll** [comment]
Anything more to debate?
